### PR TITLE
Create CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by budimlic**
Adding a file that describes the procedure and guidelines for creating pull requests. A link to this file will be automatically generated by GitHub and shown to anyone creating a pull request.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (DaoWen) [inline CONTRIBUTING.md]: Maybe we should concretely define "core team" here by creating a `habanero-rice/hclib-core` team and linking to it here.

File: CONTRIBUTING.md
Comment (DaoWen) [inline CONTRIBUTING.md]: I think we should delete everything in this sentence after the comma. Once sensitive data is pushed to GitHub, it's compromised, so just reverting the change is insufficient. I can see how it might still be a good idea to do some damage control, but in that case you'd need to force-push directly to master, not open a pull request. Pull requests stay around forever (they're impossible to delete), so if you open a pull request to erase sensitive data, the sensitive data is still visible in the pull request. But yeah, I don't see a compelling reason to include this type of exception here.

**Review by DaoWen — CHANGES_REQUESTED** [review]
This looks pretty good overall, but I think we should delete the exception about merging your own PRs.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: The section on pull requests should be above the one on commits. I'm not even sure we need a section on commits, as there could just be a paragraph in the section on pull requests talking about commit messages. This guideline of "do only one thing per commit" should really be "do only one thing per pull request", so while it isn't wrong it also doesn't capture the whole spirit of this guideline.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: In general, our commits do not have subject lines and a body. I think copy-pasting these guidelines goes against the existing state of the repo, and in general I disagree with several of them (or think they are unnecessarily restrictive). We'd have to have a discussion if we wanted all of our commit messages to start following these principles.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: Rather than just pointing to external resources, I think we should heavily modify and include these guidelines inline. What if the external resource changes? What if the link breaks? What if some of the things they say don't really apply? For example, this linked resource says "Add tests". That's a pretty useless guideline if we don't also tell people how to add tests, where to put them, etc.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: How does an external committer know 1) who the core team is, and 2) other contributors who may be affected by your changes?

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: Again, external resources shouldn't be used to define core operating principles.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: How do the core members coordinate on consensus? How is it enforced that something isn't merged before everyone agrees? Again, how do we determine who a review is blocked on?

File: CONTRIBUTING.md
Comment (agrippa) [reply CONTRIBUTING.md]: Agreed. Pushing to master is a privilege, not a right. It should not be abused, nor should any changes to master (pushes, reverts, etc) be done unilaterally.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: Guidelines for including tests in new pull requests?

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: Should we ask for each pull request to be squashed to a single commit, in the spirit of making small, incremental, separable changes? Which would also make this advice unnecessary.

**Review by agrippa — COMMENTED** [review]
What about guidelines for reviewers?

Guidelines for when things should be reverted vs. when a follow-on pull request should be requested?

File: CONTRIBUTING.md
Comment (budimlic) [inline CONTRIBUTING.md]: OK, so I'll delete the section on commits and adding a paragraph with commit guidelines to the section on pull request. I still think there should still be some commit message guidelines, I like 1., 5. and 7.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: OK, so I'll delete the section on commits and adding a paragraph with commit guidelines to the section on pull request. I still think there should still be some commit message guidelines, I like 1., 5. and 7.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: I removed Shams from @habanero-rice/habanero-rice-devs and added Nick. I think that should be the Core team (with perhaps moving Vivek out of it and leave him as an Owner). Everyone else is an Ouside collaborator.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: OK, agreed.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: Will link to @habanero-rice/habanero-rice-devs as the core team. Checking who previously committed the code they are modifying is a good indication to who might be interested/affected by what they are adding.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: Fair enough. I'll import and rewrite these guidelines.

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: Well, that's a tough one. Asking for a formal vote/consensus of all core member does seem like an overkill. On the other hand, I do want to make sure everyone on the core team has had a chance to take a look at the pull request and comment. Suggestions are welcome here. Maybe a  time period after which any core member can perform a merge if all the core members that did comment on the pull request seem to be satisfied with the state of it?

File: CONTRIBUTING.md
Comment (budimlic) [reply CONTRIBUTING.md]: Sure. If people start squashing their unrelated commits just to comply with this request, we can always ask them to separate their changes when they submit the pull request

File: CONTRIBUTING.md
Comment (agrippa) [reply CONTRIBUTING.md]: Could you show me an example of the type of commit you're thinking of? For example, at the moment I don't believe a single commit in this repo has both a subject and a body, so that would be a pretty disruptive change to everyone's development practices today. That doesn't mean it's a bad idea, just something that's definitely going to need more discussion. In general, most repos don't have subject and body, but if it's something you feel strongly about we can definitely talk about adding that requirement.

File: CONTRIBUTING.md
Comment (agrippa) [reply CONTRIBUTING.md]: Okay, I agree with adding the first part to the text. But the second part seems like it should just be the responsibility of the "core team" reviewing the code to judge the disruptiveness of the change and decide if any special notification is required? We want to have a rigorous process but also not an onerous one, otherwise we de-incentivize contributions. Asking someone to figure out everyone who's committed to a file even in just the last month sounds painful to me, and is likely to turn others away.

File: CONTRIBUTING.md
Comment (agrippa) [reply CONTRIBUTING.md]: In past jobs, one convention I've worked with is that reviews are only integrated after 2 reviewers sign off, one of which is a "gatekeeper". A "gatekeeper" is basically an expert, in this case analogous to what we're calling a member of the "core team". Now, we don't really have non-gatekeepers at the moment; pretty much everyone pushing to this repo right now would qualify as one. So we could just say that two members of the core team are required to sign off on each review. After the first completes their review and all comments are addressed they simply leave a comment or something indicating their approval. Once the second has similarly completed a review and had all comments addressed, they do the actual merge.

The only complication for our present state is that Nick and I are the most active in terms of code changes and reviews (though most of my pushes aren't going into master), so to have a third reviewer who isn't the person putting out the review would require you (Zoran) to be responsive to regular review requests. However, if the review guidelines are followed then the result should be that those reviews don't consume large amounts of time.

File: CONTRIBUTING.md
Comment (agrippa) [reply CONTRIBUTING.md]: Agreed. Things shouldn't be squashed for convenience, just for simplifying the log.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: It looks like here we're using two words (pull requests + patches) to mean the same thing. Can we just always refer to it as a pull request?

Comment (agrippa) [reply CONTRIBUTING.md]: e.g. below it says "A complex title or summary is a good indication that your pull request is doing too many things, and should be split into multiple patches" when it seems like it would be simpler to just say "A complex title or summary is a good indication that your pull request is doing too many things, and should be split into multiple pull requests"

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: oxford comma? "suggestions, concerns, and critiques"

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: "have merge privileges", delete "the"

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md]: "they may perform the actual merge", insert "may"

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md:156]: Inconsistent capitalization (or not capitalization) of pull request. I'm fine with either capitalizing or not, just want it consistent.

File: CONTRIBUTING.md
Comment (agrippa) [inline CONTRIBUTING.md:164]: I would add that we encourage any newly merged code to be no wider than 80 columns.

**Review by agrippa — COMMENTED** [review]
Looks great! I think with just a few small changes this'll be good to go and then Nick can merge it once he finishes his review. I'm sure we'll run into cases not covered by this doc, but we can tackle those issues when we come to them. Thanks for doing this!

**Review by DaoWen — APPROVED** [review]

**Comment by agrippa** [comment]
This'll need to get squashed to one commit before merging (as stated in the new contributing guidelines!)

**Comment by budimlic** [comment]
I'll squash all the changes into a single commit once Max approves the changes

**Comment by agrippa** [comment]
Yep, looks good. Just let me know when it's ready and I'll merge it.

**Comment by budimlic** [comment]
Done!

**Comment by DaoWen** [comment]
👍  In the spirit of keeping the log simple, I'm going to do the "Squash and merge" operation rather than the default merge operation. That way we won't have so many pull request merge commits cluttering our commit log.

**Pull request merged**
